qcodes/dataset/param_spec.py

```diff
@@ -61,45 +69,82 @@ def depends_on(self):
         """
         self._inferred_from.extend(
             p.name if isinstance(p, ParamSpec) else p
-            for p in inferred_from
-        )
+            for p in inferred_from)
 
-    def add_depends_on(
-            self,
-            depends_on: Sequence[Union['ParamSpec', str]]
-    ) -> None:
+    def add_depends_on(self,
+                       depends_on: Sequence[Union['ParamSpec', str]]) -> None:
         """
         Args:
             depends_on: the parameters that this parameter depends on
         """
         self._depends_on.extend(
             p.name if isinstance(p, ParamSpec) else p
-            for p in depends_on
-        )
+            for p in depends_on)
 
     def copy(self) -> 'ParamSpec':
         """
         Make a copy of self
         """
-        return ParamSpec(
-            self.name,
-            self.type,
-            self.label,
-            self.unit,
-            self._inferred_from,
-            self._depends_on
-        )
+        return ParamSpec(self.name, self.type, self.label, self.unit,
+                         deepcopy(self._inferred_from),
+                         deepcopy(self._depends_on))
 
     def sql_repr(self):
         return f"{self.name} {self.type}"
 
     def __repr__(self):
-        return f"{self.name} ({self.type})"
+        return (f"ParamSpec('{self.name}', '{self.type}', '{self.label}', "
+                f"'{self.unit}', inferred_from={self._inferred_from}, "
+                f"depends_on={self._depends_on})")
+
+    def serialize(self) -> Dict[str, Any]:
+        """
+        Write the ParamSpec as a dictionary
+        """
+        output: Dict[str, Any] = {}
+        output['name'] = self.name
+        output['paramtype'] = self.type
+        output['label'] = self.label
+        output['unit'] = self.unit
+        output['inferred_from'] = self._inferred_from
+        output['depends_on'] = self._depends_on
+        output['guids'] = self.guids
+        output['version'] = self.version
+
+        return output
+
+    @classmethod
+    def deserialize(cls, ser: Dict[str, Any]) -> 'ParamSpec':
+        """
+        Create a ParamSpec instance of the current version
+        from a serialized ParamSpec of some version
+
+        The version changes must be implemented as a series of transformations
+        of the serialized dict.
+        """
+
+        def version_0_to_1(ser):
+            if ser['version'] > 0:
+                return
+            else:
+                ser['guids'] = []
+
+        upgrades = [version_0_to_1]
+
+        if cls.version > ser['version']:
+            for upgrade in upgrades[:cls.version]:
+                upgrade(ser)
+
+        return ParamSpec(name=ser['name'],
+                         paramtype=ser['paramtype'],
+                         label=ser['label'],
+                         unit=ser['unit'],
+                         inferred_from=ser['inferred_from'],
+                         depends_on=ser['depends_on'])
```

Comment:
should here also be `guids=ser['guids']` kwarg?

Reply:
Miss me with those guids in the paramspec